### ISSUE TITLE
Cause refresh of asyncCallback options

### ISSUE THREAD

**Issue by TrevorPage** [OptimalBPM/angular-schema-form-dynamic-select#23]
My form contains several selects that make use of `asyncCallback`. Example extract:

```
            {
                "key" : "CONTACT_ID",
                "type": "strapselect",
                "options": {
                    "asyncCallback": MessageRules.fetchContactOptions
                },
                "onChange" : "contactIdChanged()"
            },
            {
                "key" : "PHONE_INDEX",
                "type": "strapselect",
                "options": {
                    "asyncCallback": MessageRules.fetchContactTelephoneOptions
                },
                "condition" : (.... not relevant ....)
            },
```

The asynchronous callbacks for both of the above example selects make HTTP GET calls to get the name/value data. (I appreciate I could use `httpGet` to much simplify this, but I have a problem achieving necessary transformation using `map`, which I might raise in a separate issue tracker.)

The HTTP call that is made for the second select depends on the selection of the first select. For that reason, I would like the second select's `asyncCallback` to fire again whenever the first select is changed. I have attempted to cause this using the `onChange` in the first select as you can see, which broadcasts a `schemaFormRedraw`:

```
    $scope.contactIdChanged = function() {
       $timeout(function() { 
        $scope.$broadcast('schemaFormRedraw'); 
        }, 0);
    }  
```

This does not appear to cause ASFDS to call the `asyncCallback`s to repopulate the maps. Furthermore in the console I quite often see (for example):

`dynamicSelectController.populateTitleMap(key:PHONE_INDEX) : There is already a titleMap`

It could be my imagination but I _thought_ I had this working at one point, so I'm not sure if any change in the library or what I'm doing has broken it. 

Is there a way I can cause the selects to refresh using a fresh GET please?

Thank you :)

**Comment by TrevorPage** [comment]
I now realise that after the async callback has done its stuff, each array element in the $scope.form object has a `titleMap` added. Once the `titleMap` is present in the `form`, ASFDS won't bother re-firing the callback when ASF is made to refresh. 

Solutions that work for me are to either:
- Just re-initialise the `$scope.form` again, to get rid of the `titleMaps`. Then immediately follow with `$scope.$broadcast('schemaFormRedraw')`. Or,
- Loop over the `$scope.form` elements and `delete` the `titleMap` properties, then `$scope.$broadcast('schemaFormRedraw')`. This is possibly a bit better as it would retain any other state that ASF/DS puts in there. 

Am I missing a blindingly obvious better way though?

Edit: While the above solutions do the job, the obvious refresh of the form does not look pretty (because you can see the selects momentarily moving to the first option and then back again to the selected option). I suppose what would be nice would be a way to trigger a specific select to re-populate.

**Comment by nicklasb** [comment]
I am not _completely_ sure I understand. But as it is all async, would a post fetch-event be of use?
In other news, I normally load _all_ the possible values in the second select and use the [filter](https://github.com/OptimalBPM/angular-schema-form-dynamic-select#filters) functionality instead.

**Comment by TrevorPage** [comment]
To try to explain it a little further, imagine that the first select is used to choose a person, and the second select is used to choose one of many telephone numbers for that person. The requirement is that when the person select is changed, the telephone number select is to be re-populated with the numbers for that person. Retrieval of telephone numbers is done on a per-person basis, by providing the person ID to the REST endpoint.

This is quite a contrived/simplified example compared with the real use case. I suppose I could do it so that in the background all numbers for all users are fetched from the server, but I would prefer not to. 

I think this use case is rather similar to another example I saw someone mention once (either in an issue on this library, or on ASF) where for one select you choose a country, and for the second select you choose a state/county within that country. Each time the country select is changed, you'd want the facility to make the region select re-populate with values from the server.

I have this working as I want now, but it's achieved by re-assigning or modifying the form, which isn't entirely pretty from a coding perspective but it works.

I'm not sure exactly how ASFDS would do it, but one I idea is to have a function I can call on the form object or something, that I'd supply the key for the control I want to repopulate. E.g. `form.refreshTitleMap("PHONE_INDEX")`.

**Comment by nicklasb** [comment]
Ok. I see.
I am not much for a complete redraw, I'd rather add some way to easily tell a field to refresh its titleMap.
Wouldn't a simple lookup function or .refreshTitleMap(key) help?

**Comment by TrevorPage** [comment]
I reckon the .refreshTitleMap(key) would be great!

**Comment by nicklasb** [comment]
Haha, I didn't see that you had actually proposed the exact form.refreshTitleMap() function, I misread that. Sorry, didn't want to steal your thunder. :-)

Anyway, I have thought about it, and I am not completely sure what is the correct way angular-wise to make each controller scope available to the app.js context. Or even if one is supposed to. 

But one way could be to, like with the title map, add a property to each form with the scope of each controller.
That way one could, from the application scope, in onChange, call $scope.form[2].options.scope.populateTitleMap($scope.form[2]) 

A form.refreshTitleMap(key) is harder as each instance of ASFDS doesn't know about the others, that would just result in a lot of scope-tree-traversal which might end up not working some ASF change later.

**Comment by nicklasb** [comment]
The only bad thing about it is that you'd have to lookup what place in the form array the "PHONE_INDEX"-key in the array has. But that is trivial, and also, access to the scope basically gives you total control over ASFDS:s behavior.

**Comment by nicklasb** [comment]
If you now look at the develop branch, I have added the options.scope property, it is set to the scope of the ASFDS component. I have also added to the examples, a somewhat lame, as nothing visible happens, but still an example, on how it can be used in the multiselectDynamicAsync's onChange event.

**Comment by TrevorPage** [comment]
Hi Nicklas, 

That's absolutely fantastic! Thank you for doing this. I'm hoping to give it a test before or during the weekend.

Trev

**Comment by nicklasb** [comment]
Good, I hope it works out for you.

//Nicklas

**Comment by naterkane** [comment]
I'm faced with a very similar situation, where one select's value is dependent on another. though I can (and am in the interim trying) to use filter on the second select, i would really prefer to onChange of the first select's value, request a new set from the api to populate the second, dependent select. 
- @TrevorPage how did things work out for you? 
- @nicklasb which commit of the dev branch did you add this? I really don't want to use that branch, so I'd like to fix it to a specific commit if possible.

also, once this scenario is resolved, can someone update the docs?

**Comment by TrevorPage** [comment]
Apologies - I haven't tried it yet.

**Comment by naterkane** [comment]
well, i'm going to be trying it this morning. 

and the commit is `#6281c8548c533eb8ff726d61c67af21b8494770d`, which is most recent under #develop

**Comment by nicklasb** [comment]
Hi, everyone.

@naterkane: 

Actually, no...the relevant commit is #fc45fa1f19cdc3b34cec082597a50f80f3a07d52 as far as I can see?

WRT the docs, I am not sure what you mean..the master branch docs will not be updated until the develop branch is merged. And either way, please be a little more patient, you are commenting on a single commit on the develop branch, documentation might sometimes have to wait until a feature is fully implemented.

In this case, however, the feature was all in one commit, so the documentation got updated in the next commit - the #6281c8548c533eb8ff726d61c67af21b8494770d - commit.

**Comment by naterkane** [comment]
I see that now. thanks!

**Comment by nicklasb** [comment]
Did it work out?

**Comment by naterkane** [comment]
I've actually been held up at the moment taking care of some other
things... it's on a punch list for later today though.

On Tue, Jun 23, 2015 at 12:39 PM, Nicklas Börjesson <
notifications@github.com> wrote:

> Did it work out?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OptimalBPM/angular-schema-form-dynamic-select/issues/23#issuecomment-114566228
> .

## 

Best,
Nater Kane

**Comment by nicklasb** [comment]
@naterkane : Gentle reminder.. Have you tried it?

**Comment by nicklasb** [comment]
@TrevorPage: Asking you as well.
It would be great if you guys could try it.

**Comment by TrevorPage** [comment]
Apologies again - I will most definitely test this as soon as I can. Thanks again for your efforts.

**Comment by TrevorPage** [comment]
Still not forgotten.

**Comment by nicklasb** [comment]
No problem, just thinking of wrapping up another release.

**Comment by nicklasb** [comment]
@TrevorPage, @naterkane : Ok, I need to wrap up this release now, I see no problems, it is an uncomplicated addition, and it doesn't seem to add any secondary effects, so I'll add it.

**Comment by vincetastic** [comment]
Late to the party here but, unfortunately, the addition of the scope to the form creates an error. 

The problem is that in response to the broadcast of "schemaFormRedraw", the scope.$on('schemaFormRedraw') function in schema-form.js is called. It then attempts to deep copy the form, which now includes the scope as a result of the fix for this issue.

However, Angular throws the following error on attempt to copy a scope: 

Error: [ng:cpws] Can't copy! Making copies of Window or Scope instances is not supported.

I've worked around this by modifying the schema-form.js code to detect the scope in any options object of any form properties and nullifying it before copying, then restoring it to the resulting form after copy.

So, from this:

```
      var form   = scope.initialForm ? angular.copy(scope.initialForm) : ['*'];
```

To this:

```
      var savedScopes = {};
      if (scope.initialForm) {
          for (var prop in scope.initialForm) {
              if (scope.initialForm[prop].options && scope.initialForm[prop].options.scope) {
                  savedScopes[prop] = scope.initialForm[prop].options.scope;
                  scope.initialForm[prop].options.scope = null;
              }
          }
      }

      var form   = scope.initialForm ? angular.copy(scope.initialForm) : ['*'];

      for (var prop in savedScopes) {
          form[prop].options.scope = savedScopes[prop];
      }
```

**Comment by nicklasb** [comment]
Strange. I am not getting this problem? Is this when you explicitly trigger schemaFormRedraw?

**Comment by vincetastic** [comment]
Yes. The redraw function that's triggered by triggering schemaFormRedraw immediately executes the line to copy the form, whereupon I see the error thrown.

You may need to check the error console for messages. 

Else, maybe it's down to angular version? Not sure when they added this check to angular.copy(), but I'm on 1.5.8.

**Comment by gatperdut** [comment]
updating from angular 1.5.0 to 1.5.8 fixed this for me ([ng:cpws] Can't copy! Making copies of Window or Scope instances is not supported.)

**Comment by gatperdut** [comment]
I'm getting the can't copy scope error. Any updates on this, or a workaround?

**Comment by nicklasb** [comment]
If would be great to have a plunkr that displays the problem.

**Comment by anasofiapaixao** [comment]
I have made one: https://next.plnkr.co/plunk/Hp3XwxInVxhstiuW

**Comment by velteyn** [comment]
the exact same problem happens on the official example . If you put a $scope.$broadcast('schemaFormRedraw');
at the end of anything the error shows up !!

**Comment by nicklasb** [comment]
Then it should not be related to this plugin, but ASF?

**Comment by velteyn** [comment]
I'm using ASF normally without your plugin and I build a form dynamically in response to data arriving from a rest service.
I have to use $scope.$broadcast('schemaFormRedraw'); massivelly to redraw anyting. When they ask me to put also a UI-SELECT I tried to implement your plugin. The piker works. But the refresh and the dynamik building is broken by the error triggered by refresh:
angular.js:11655 Error: [ng:cpws] http://errors.angularjs.org/1.3.15/ng/cpws
    at http://localhost:8080/sanmarco-informatica-theme/js/vendor/angular.min.js:6:417

**Comment by nicklasb** [comment]
I am not sure what you mean, are you saying there is a problem with ASFDS or not?

**Comment by velteyn** [comment]
Sorry, I'll try to be more clear .. I use ASF without problems.. when I used ASFDS I incurred into the problem described into the subject. So yes ,according to me there is a problem in ASFDS.

**Comment by nicklasb** [comment]
Yes. But if the same problem occurs in ASF, then the problem might still be in ASF, but only _triggered_ by you using ASFDS. 
The problem seems to be that you have to call $scope.$broadcast('schemaFormRedraw') ? 
@vincetastic 's reply above seems to indicate that a fix in schema-form.js is needed?

**Comment by velteyn** [comment]
Dunno !! But I tried to apply @vincetastic solution but in my case the problem is  still there... I had to remove your plugin and gave up to UI-SELECT :( Unfortunatelly I have no time for a deep investigation
